api/streams.go

```diff
@@ -0,0 +1,254 @@
+
+	// Start go routine for receiving messages from the stream (especially relevant for bi-directional streams).
+	go c.recvLoop(s)
+
+	// Start go routine for sending messages from the channel to the stream
+	go c.sendLoop(s)
+
+	return c, nil
+}
+
+// removeStream deletes the channel from the stream map.
+func (s *StreamsOf[StreamType, MsgType]) removeStream(target string) {
+	s.log.Debugf("Removing stream channel for target %s", target)
+
+	s.mutex.Lock()
+	delete(s.channels, target)
+	s.mutex.Unlock()
+}
+
+// sendLoop continuously fetches new messages from the channel inside c and sends them to the appropriate stream.
+func (c *StreamChannelOf[StreamType, MsgType]) sendLoop(s *StreamsOf[StreamType, MsgType]) {
+	var err error
+
+	// Fetch new messages from channel (this will block)
+	for m := range c.channel {
+		// Try to send the message in our stream
+		err = c.stream.SendMsg(m)
+		if errors.Is(err, io.EOF) {
+			s.log.Infof("Stream to %s (%s) closed with EOF", c.component, c.target)
+
+			// Remove the stream from our map and end this goroutine
+			s.removeStream(c.target)
+			return
+		}
+
+		// Some other error than EOF occurred
+		if err != nil {
+			s.log.Errorf("Error when sending message to %s (%s): %v", c.component, c.target, err)
+
+			// Close the stream gracefully. We can ignore any error resulting from the close here
+			_ = c.stream.CloseSend()
+
+			// Remove the stream from our map and end this goroutine
+			s.removeStream(c.target)
+			return
+		}
+
+		s.log.Debugf("Message sent to %s (%s)", c.component, c.target)
+	}
+}
+
+// recvLoop continuously receives message from the stream. Currently, they are just discarded. In the future, we might
+// want to send them back to the caller. But we need to receive them, otherwise the buffer of the stream gets congested.
+func (c *StreamChannelOf[StreamType, MsgType]) recvLoop(s *StreamsOf[StreamType, MsgType]) {
+	for {
+		// TODO(oxisto): Check, if this also works for uni-directional streams
+		msg := new(dumbProtoMesssage)
+		err := c.stream.RecvMsg(msg)
+
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		if err != nil {
+			s.log.Errorf("Error receiving response from stream: %v", err)
+			break
+		}
+	}
+}
+
+// Send sends the message into the stream via the channel. Since this uses the receive operator on the channel,
+// this function may block until the message is received on the sendLoop of this StreamChannelOf or if
+// the buffer of the channel is full.
+func (c *StreamChannelOf[StreamType, MsgType]) Send(msg MsgType) {
+	c.channel <- msg
+}
+
+// defaultLog returns the default logger, if none is specified.
+func defaultLog() *logrus.Entry {
+	return logrus.NewEntry(logrus.StandardLogger())
+}
+
+// dumbProtoMesssage is used for now to give a correctly typed message to RecvMsg in recvLoop. In the future, use
+// types of response message of respective RPCs. It "implements" the protoiface.MessageV1 interface
+type dumbProtoMesssage struct{}
```

Comment:
I would suggest `emptyProtoMessage` or can't we just use emptypp.Empty?

Reply:
Sorry, I missed that. I just tested it, emptypp.Empty works too. I open a new MR for that.